cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -101,25 +126,108 @@ private void refreshCredentials() throws IOException {
             .build();
 
     StsTokenExchangeResponse response = handler.exchangeToken();
-    this.accessBoundarySessionKey = response.getAccessBoundarySessionKey();
-    this.intermediateAccessToken = response.getAccessToken();
 
-    // The STS endpoint will only return the expiration time for the intermediary token
+    synchronized (refreshLock) {
```

Comment:
qq, why do we need a lock here?

Reply:
From [comment thread](https://docs.google.com/document/d/1YLYAlae_AAOtb9Mw-TRawi5df_8EfHv-hO04HgZgiYA/edit?disco=AAABYwpOifM) in the design doc:

> We still need to synchronize access to the intermediary token and session key. Concurrent access by the background refresh task and main user thread could lead to inconsistencies. Because the background refresh task updates these values separately, there's a chance the main thread could grab a new token with an old key, causing problems.